pkg/compact/planner.go

```diff
@@ -7,47 +7,69 @@ import (
 	"context"
 
 	"github.com/go-kit/kit/log"
+	"github.com/oklog/ulid"
 	"github.com/thanos-io/thanos/pkg/block/metadata"
 )
 
 type tsdbBasedPlanner struct {
 	logger log.Logger
 
 	ranges []int64
+
+	noCompBlocksFunc func() map[ulid.ULID]*metadata.NoCompactMark
 }
 
 var _ Planner = &tsdbBasedPlanner{}
 
-// NewTSDBBasedPlanner is planner with the same functionality as Prometheus' TSDB plus special handling of excluded blocks.
+// NewTSDBBasedPlanner is planner with the same functionality as Prometheus' TSDB.
 // TODO(bwplotka): Consider upstreaming this to Prometheus.
-// It's the same functionality just without accessing filesystem, and special handling of excluded blocks.
+// It's the same functionality just without accessing filesystem.
 func NewTSDBBasedPlanner(logger log.Logger, ranges []int64) *tsdbBasedPlanner {
-	return &tsdbBasedPlanner{logger: logger, ranges: ranges}
+	return &tsdbBasedPlanner{logger: logger, ranges: ranges, noCompBlocksFunc: func() map[ulid.ULID]*metadata.NoCompactMark { return make(map[ulid.ULID]*metadata.NoCompactMark) }}
+}
+
+// NewPlanner is a default Thanos planner with the same functionality as Prometheus' TSDB plus special handling of excluded blocks.
+// It's the same functionality just without accessing filesystem, and special handling of excluded blocks.
+func NewPlanner(logger log.Logger, ranges []int64, noCompBlocks *GatherNoCompactionMarkFilter) *tsdbBasedPlanner {
+	return &tsdbBasedPlanner{logger: logger, ranges: ranges, noCompBlocksFunc: noCompBlocks.NoCompactMarkedBlocks}
 }
 
+// TODO(bwplotka): Consider smarter algorithm, this prefers smaller iterative compactions vs big single one: https://github.com/thanos-io/thanos/issues/3405
 func (p *tsdbBasedPlanner) Plan(_ context.Context, metasByMinTime []*metadata.Meta) ([]*metadata.Meta, error) {
-	res := selectOverlappingMetas(metasByMinTime)
+	noCompactMarked := p.noCompBlocksFunc()
+	notExcludedMetasByMinTime := make([]*metadata.Meta, 0, len(metasByMinTime))
+	for _, meta := range metasByMinTime {
+		if _, excluded := noCompactMarked[meta.ULID]; excluded {
+			continue
+		}
+		notExcludedMetasByMinTime = append(notExcludedMetasByMinTime, meta)
+	}
+
+	res := selectOverlappingMetas(notExcludedMetasByMinTime)
 	if len(res) > 0 {
 		return res, nil
 	}
-
 	// No overlapping blocks, do compaction the usual way.
-	// We do not include a recently created block with max(minTime), so the block which was just created from WAL.
-	// This gives users a window of a full block size to piece-wise backup new data without having to care about data overlap.
+
+	// We do not include a recently created block with max(minTime), so the block which was just uploaded to bucket.
```

Comment:
Why a block with the max `minTime` is just uploaded?  It is possible that it was updated several days ago, but has the max `minTime`

Reply:
This is exactly what this comment says right?  That it has max of minTimes of all blocks, no?

Reply:
For example, if we stop adding new blocks to a bucket, and the latest block in the bucket was uploaded 3 days ago. So we also want to exclude the latest block in this case?

The comment mentions `recently created block` so seems to check the block creation time makes more sense. Maybe I miss something

Reply:
We mean the freshest TBH

also the whole check is really not super needed for us. It's relevant to Prometheus, but let's keep it :hugs:

Reply:
Ok. If it is relevant to Prometheus then it makes sense :+1: